pkg/pool-hooks/contracts/StableSurgeHookExample.sol

```diff
@@ -0,0 +1,243 @@
+     * @param pool The pool on which the hook was registered
+     */
+    event StableSurgeHookExampleRegistered(address indexed pool);
+
+    /**
+     * @notice The threshold percentage has been changed for a pool in a `StableSurgeHookExample` contract.
+     * @dev Note, the initial threshold percentage is set on deployment and an event is emitted.
+     * @param pool The pool for which the threshold percentage has been changed
+     * @param newSurgeThresholdPercentage The new threshold percentage
+     */
+    event ThresholdSurgePercentageChanged(address indexed pool, uint256 newSurgeThresholdPercentage);
+
+    /// @notice The sender does not have permission to call a function.
+    error SenderNotAllowed();
+
+    /// @notice The threshold must be a valid percentage value.
+    error InvalidSurgeThresholdPercentage();
+
+    constructor(IVault vault, uint256 defaultSurgeThresholdPercentage) VaultGuard(vault) {
+        _ensureValidPercentage(defaultSurgeThresholdPercentage);
+
+        _defaultSurgeThresholdPercentage = defaultSurgeThresholdPercentage;
+    }
+
+    /// @inheritdoc IHooks
+    function getHookFlags() public pure override returns (HookFlags memory hookFlags) {
+        hookFlags.shouldCallComputeDynamicSwapFee = true;
+    }
+
+    /**
+     * @notice Getter for the default surge threshold percentage.
+     * @return surgeThresholdPercentage The default surge threshold percentage for this hook contract
+     */
+    function getDefaultSurgeThresholdPercentage() external view returns (uint256) {
+        return _defaultSurgeThresholdPercentage;
+    }
+
+    /**
+     * @notice Getter for the surge threshold percentage for a pool.
+     * @param pool The pool for which the surge threshold percentage is requested
+     * @return surgeThresholdPercentage The surge threshold percentage for the pool
+     */
+    function getSurgeThresholdPercentage(address pool) external view returns (uint256) {
+        return _surgeThresholdPercentage[pool];
+    }
+
+    /// @inheritdoc IHooks
+    function onRegister(
+        address,
+        address pool,
+        TokenConfig[] memory,
+        LiquidityManagement calldata
+    ) public override onlyVault returns (bool) {
+        emit StableSurgeHookExampleRegistered(pool);
+
+        // Initially set the pool threshold to the default (can be changed by the pool swapFeeManager in the future).
+        _setSurgeThresholdPercentage(pool, _defaultSurgeThresholdPercentage);
+
+        return true;
+    }
+
+    /// @inheritdoc IHooks
+    function onComputeDynamicSwapFeePercentage(
+        PoolSwapParams calldata params,
+        address pool,
+        uint256 staticSwapFeePercentage
+    ) public view override onlyVault returns (bool, uint256) {
+        uint256 amp;
+        (amp, , ) = IStablePool(pool).getAmplificationParameter();
+
+        // In order to calculate `weightAfterSwap` we need balances after swap, so we must compute the swap amount.
```

Comment:
Is this absolutely needed? Can't we do an approximation with the current balances and the amounts given?

Sounds a bit overkill to compute the whole swap just to compute the fee percentage, given that we'll need to do it all over again in the pool `onSwap` function right after this.

Intuitively, we know whether we start in a balanced / unbalanced state, and we can sort of guess whether the swap will push it further in the right direction or not. Although I get it might be more nuanced than that, as you might start in an imbalanced state, the direction might be correct but the magnitude of the swap pushes it into unbalanced territory but on the opposite side.

Reply:
If there's really no way around it, why not just call `pool.onSwap` and get a result as opposed to reimplementing all the logic here?